npm/private/npm_translate_lock.bzl

```diff
@@ -456,183 +488,531 @@ def _check_for_conflicting_public_links(npm_imports, public_hoist_packages):
-                content = rctx.read(rctx.attr.npm_package_lock),
-                executable = False,
-            )
-        elif rctx.attr.yarn_lock:
-            lock_attr = rctx.attr.yarn_lock
-            rctx.file(
-                "yarn.lock",
-                content = rctx.read(rctx.attr.yarn_lock),
-                executable = False,
-            )
-        else:
-            fail("rules_js internal validation error, please file an issue")
+    msg = """
 
-        if rctx.attr.npmrc:
-            rctx.file(
-                ".npmrc",
-                content = rctx.read(rctx.attr.npmrc),
-                executable = False,
-            )
+INFO: Initial pnpm-lock.yaml file generated. Please add the generated pnpm-lock.yaml file into
+      source control and set the pnpm_lock attribute in npm_translate_lock(name = '{name}') to '{label}'
+      and then run your build again.""".format(
+        name = rctx.name,
+        label = expected_pnpm_lock_label,
+    )
+    fail(msg)
+
+def _update_data_hashes_file(rctx, translate_state):
+    contents = [
+        "# Generated by {}".format(translate_state["rule_description"]),
+    ]
+    for key, value in translate_state["input_hashes"].items():
+        contents.append("{}={}".format(key, value))
+    data_hashes_file = paths.join(paths.dirname(paths.join(translate_state["pnpm_lock"].package, translate_state["pnpm_lock"].name)), PNPM_LOCK_INPUT_HASHES_FILENAME)
+    rctx.file(data_hashes_file, "\n".join(contents) + "\n")
+    _reverse_force_copy(
+        rctx,
+        _sibling_label(translate_state["pnpm_lock"], PNPM_LOCK_INPUT_HASHES_FILENAME),
+        _sibling_path(rctx, translate_state["pnpm_lock"], PNPM_LOCK_INPUT_HASHES_FILENAME),
+    )
 
-        result = rctx.execute([
+def _dicts_match(a, b):
+    if len(a) != len(b):
+        return False
+    for key in a.keys():
+        if a[key] != b[key]:
+            return False
+    return True
+
+def _update_pnpm_lock(rctx, translate_state):
+    if not translate_state["pnpm_lock"]:
+        fail(INTERNAL_ERROR_MSG)
+
+    expected_input_hashes_path = _sibling_path(rctx, translate_state["pnpm_lock"], PNPM_LOCK_INPUT_HASHES_FILENAME)
+    expected_pnpm_lock_label = _sibling_label(translate_state["pnpm_lock"], PNPM_LOCK_INPUT_HASHES_FILENAME)
+    if _exists(rctx, expected_input_hashes_path):
+        input_hashes = parse_npmrc(rctx.read(expected_pnpm_lock_label))
+        if _dicts_match(input_hashes, translate_state["input_hashes"]):
+            # Calculated input hashes match saved input hashes; nothing to update
+            return False
+
+    if RULES_JS_FROZEN_PNPM_LOCK_ENV in rctx.os.environ.keys() and rctx.os.environ[RULES_JS_FROZEN_PNPM_LOCK_ENV]:
+        fail("""
+
+ERROR: {pnpm_lock} file's inputs have changed since the last update and an update may be necessary.
+       Run `bazel query @{name}//...` locally to apply updates.""".format(pnpm_lock = translate_state["pnpm_lock"], name = rctx.name))
```

Comment:
What's a better command to run than `bazel query @npm//...`?

Reply:
`bazel fetch @npm//...` does the wrong thing. I wish there was a `bazel sync <repo>` command.